tests/abi/tools/revng-check-compatibility-with-abi/Verify.cpp

```diff
@@ -230,114 +227,299 @@ void VH::verify(const abi::FunctionType::Layout &FunctionLayout,
+  auto LookupHelper = [&Helper](const auto &Test) {
+    return Test.Name == Helper.FunctionName;
+  };
 
-  std::size_t TestedCount = 0;
+  std::size_t ArgumentTestCount = 0, ReturnValueTestCount = 0;
   for (auto &Function : Binary->Functions()) {
-    const std::string &Name = Function.OriginalName();
-    auto CurrentFunction = ParsedArtifact.Functions.find(Name);
-    if (CurrentFunction == ParsedArtifact.Functions.end()) {
-      // Ignore types not present in the artifact.
+    Helper.FunctionName = Function.OriginalName();
+    if (Helper.FunctionName.take_front(5) == "test_")
+      Helper.FunctionName = Helper.FunctionName.drop_front(5);
+    if (auto Test = llvm::find_if(Parsed.ArgumentTests, LookupHelper);
+        Test != Parsed.ArgumentTests.end()) {
+      Helper.arguments(getPrototypeLayout(Function, ABI), *Test);
+      ++ArgumentTestCount;
+    } else if (auto Test = llvm::find_if(Parsed.ReturnValueTests, LookupHelper);
+               Test != Parsed.ReturnValueTests.end()) {
+      Helper.returnValue(getPrototypeLayout(Function, ABI), *Test);
+      ++ReturnValueTestCount;
+    } else {
+      // Ignore types from the model, that are not mentioned in
+      // the runtime test artifact.
       continue;
     }
+  }
 
-    ++TestedCount;
-    Helper.FunctionName = Name;
-
-    using Layout = abi::FunctionType::Layout;
-    const model::Type *Prototype = Function.Prototype().getConst();
-    revng_assert(Prototype != nullptr);
-    if (auto *CABI = llvm::dyn_cast<model::CABIFunctionType>(Prototype)) {
-      // Copy the prototype since we might have to modify it before testing.
-      model::CABIFunctionType PrototypeCopy = *CABI;
-
-      if (ABI != PrototypeCopy.ABI()) {
-        // Workaround for dwarf sometimes misdetecting a very specific ABI,
-        // for example, it labels `SystemV_x86_regparam_N` as just `SystemV_x86`
-        // despite them being incompatible.
-        std::array AllowedABIs = { model::ABI::SystemV_x86_regparm_3,
-                                   model::ABI::SystemV_x86_regparm_2,
-                                   model::ABI::SystemV_x86_regparm_1 };
-        revng_assert(llvm::is_contained(AllowedABIs, ABI));
-        PrototypeCopy.ABI() = ABI;
-      }
-
-      Helper.verify(Layout(PrototypeCopy), *CurrentFunction);
-    } else if (auto *Raw = llvm::dyn_cast<model::RawFunctionType>(Prototype)) {
-      Helper.verify(Layout(*Raw), *CurrentFunction);
-    } else {
-      // Ignore non-function types.
-    }
+  if (ArgumentTestCount != Parsed.ArgumentTests.size()) {
+    std::string Error = std::to_string(ArgumentTestCount)
+                        + " functions from the binary were tested, but "
+                          "artifact contains "
+                        + std::to_string(Parsed.ArgumentTests.size())
+                        + " argument test functions.\n"
+                          "Does the binary match the artifact?";
+    revng_abort(Error.c_str());
   }
```

Comment:
My main suggestions on Mattermost boils down to this: instead of hardcoding all these assertions in the code, we could emit a new file for the *actual* and *expected* data and then diff them.

Using assertions requires you to use GDB to debug stuff, while using diff you can look at the bigger picture and see all the differences altogether.

On the other hand 1) I'm not really excited at the idea of introducing yet another file format for testing purpose, 2) I really want to get this done.

Reply:
To be fair, I found that the use of GDB really reduced after I made all of these messages really explicit: In ~95% of the case looking at the error message + `yaml` dumped by the runner is enough to pinpoint the exact issue.

The only thing I'd also add to cover ~4.5% more is to add a logger here dumping function types and layouts everywhere, but maybe not today. I'll probably just do at the first instance when I find that I want to `GDB` into this.

Reply:
Because of needing to debug error I introduced today while shuffling `arguments` around, I needed to debug it 😅. So, now there's a `fail` function that dumps tested function name and layout on top of the `revng_abort` message. Please take a look.